cuda/src/utils/cuda/async_copy.cpp

```diff
@@ -26,13 +27,39 @@ namespace {
 struct cuda_event : public vecmem::abstract_event {
 
     /// Constructor with the created event.
-    cuda_event(cudaEvent_t event) : m_event(event) {}
+    cuda_event(cudaEvent_t event) : m_event(event) {
+        assert(m_event != nullptr);
+    }
     /// Destructor
-    ~cuda_event() { VECMEM_CUDA_ERROR_CHECK(cudaEventDestroy(m_event)); }
+    ~cuda_event() {
+        // Check if the user forgot to wait on this asynchronous event.
+        if (m_event != nullptr) {
+            // If so, wait implicitly now.
+            VECMEM_DEBUG_MSG(1, "Asynchronous CUDA event was not waited on!");
+            wait();
+#ifdef VECMEM_FAIL_ON_ASYNC_ERRORS
+            // If the user wants to fail on asynchronous errors, do so now.
+            std::terminate();
+#endif  // VECMEM_FAIL_ON_ASYNC_ERRORS
+        }
+    }
```

Comment:
@stephenswat, how do you feel about such a behaviour in these types? As I wrote in the PR, after some pondering, I found this as the best setup to allow debugging issues with user code. :thinking: